### PR TITLE
Save rest headers as transaction meta

### PR DESCRIPTION
**Pull request by grixa**
CC @dkorolev @mzhurovich

### CODE REVIEW

**Comment by dkorolev** [comment]
Looks good! Let me keep it unmerged for another ~28 hours to avoid possible external synchronization troubles.

**Comment by sompylasar** [comment]
Why non-GET? GET may also include the headers, e.g. for authentication.

**Comment by sompylasar** [comment]
I think some standard headers should be considered, too, e.g. "Referer".

**Comment by mzhurovich** [comment]
@sompylasar 
1) We had a number of discussions on GET method for REST and finally implemented it as `ReadOnlyTransaction`, so it can't persist any data.
2) Speaking of standard headers - good point, but I suppose iOS app doesn't send standard web headers, so we don't need this logging until we launch a web-based version of the app.

**Comment by mzhurovich** [comment]
@grixa LGTM

**Comment by sompylasar** [comment]
> 1) We had a number of discussions on GET method for REST and finally implemented it as ReadOnlyTransaction, so it can't persist any data.

Oh, I thought this is not persisted, rather used in some handlers. Misunderstood.

> 2) Speaking of standard headers - good point, but I suppose iOS app doesn't send standard web headers, so we don't need this logging until we launch a web-based version of the app.

Yes, sure, I just missed that this was for CTFO only. Sorry.

**Comment by dkorolev** [comment]
Also, keep in mind Current binaries are always run behind nginx. If some standard headers are required to be persisted, we can also proxy them through under a different name.

File: server.h [hunk 1]
```diff
@@ -59,6 +59,46 @@ static const size_t FLAGS_reports_to_ban = 10u;
...
+  struct RESTful : super_t::RESTful<HTTP_VERB, PARTICULAR_FIELD, ENTRY, KEY> {
+    using restful_super_t = super_t::RESTful<HTTP_VERB, PARTICULAR_FIELD, ENTRY, KEY>;
+    using headers_list_t = std::vector<current::net::http::Header>;
+    headers_list_t headers;
+
+    template <typename F>
+    void Enter(Request request, F&& next) {
+      const char prefix[] = "X-";
+      for (const auto& header : request.headers) {
+        if (header.header.substr(0, sizeof(prefix) - 1) == prefix)
```
Comment (dkorolev) [inline server.h:81]: =) ну ок

File: server.h [hunk 2]
```diff
@@ -59,6 +59,46 @@ static const size_t FLAGS_reports_to_ban = 10u;
...
+          headers.push_back(header);
+      }
+      this->restful_super_t::template Enter<F>(std::move(request), std::move(next));
```
Comment (dkorolev) [inline server.h:84]: `this->` is unnecessary here.

Comment (dkorolev) [inline server.h:84]: And the 2nd parameter is better as `std::forward<F>(next)`.

**Pull request merged**